projects/plugins/jetpack/modules/wordads/php/class-wordads-smart.php

```diff
@@ -0,0 +1,261 @@
+<?php
+/**
+ *  An implementation for ads served through Equativ Smart Ad Server.
+ *
+ * @package automattic/jetpack
+ */
+
+use Automattic\Jetpack\Assets;
+
+// phpcs:disable WordPress.WP.EnqueuedResources.NonEnqueuedScript
+
+require_once WORDADS_ROOT . '/php/class-wordads-array-utils.php';
+
+/**
+ * Contains all the implementation details for Smart ads
+ */
+class WordAds_Smart {
+
+	/**
+	 * The single instance of the class.
+	 *
+	 * @var WordAds_Smart
+	 */
+	protected static $instance = null;
+
+	/**
+	 * Is this an AMP request?
+	 *
+	 * @var bool
+	 */
+	private $is_amp;
+
+	/**
+	 * Current blog theme from get_stylesheet().
+	 *
+	 * @var string
+	 */
+	private $theme;
+
+	/**
+	 * Has Smart asset been enqueued?
+	 *
+	 * @var bool True if Smart asset has been enqueued.
+	 */
+	private $is_asset_enqueued = false;
+
+	/**
+	 * Toggle for inline ads.
+	 *
+	 * @var bool True if inline ads are enabled.
+	 */
+	private $is_inline_enabled;
+
+	/**
+	 * Private constructor.
+	 */
+	private function __construct() {
+	}
+
+	/**
+	 * Main Class Instance.
+	 *
+	 * Ensures only one instance of WordAds_Smart is loaded or can be loaded.
+	 *
+	 * @return WordAds_Smart
+	 */
+	public static function instance() {
+		if ( self::$instance === null ) {
+			self::$instance = new self();
+		}
+		return self::$instance;
+	}
+
+	/**
+	 * Initialize the ads.
+	 *
+	 * @param WordAds_Params $params Object containing WordAds settings.
+	 *
+	 * @return void
+	 */
+	public function init( WordAds_Params $params ) {
+		$this->is_amp            = function_exists( 'amp_is_request' ) && amp_is_request();
+		$this->theme             = get_stylesheet();
+		$this->is_inline_enabled = is_singular( 'post' ) && $params->options['wordads_inline_enabled'];
+
+		// Allow override.
+		// phpcs:disable WordPress.Security.NonceVerification.Recommended
+		if ( isset( $_GET['inline'] ) && 'true' === $_GET['inline'] ) {
+			$this->is_inline_enabled = true;
+		}
+		if ( $this->is_inline_enabled ) {
+			// Insert ads.
+			$this->insert_ads();
+		}
+	}
+
+	/**
+	 * Enqueue any front-end CSS and JS.
+	 *
+	 * @return void
+	 */
+	public function enqueue_assets() {
+
+		if ( $this->is_asset_enqueued ) {
+			return;
+		}
+
+		add_action( 'wp_head', array( $this, 'insert_config' ) );
+
+		Assets::register_script(
+			'adflow_script_loader',
+			'_inc/build/wordads/js/adflow-loader.min.js',
+			JETPACK__PLUGIN_FILE,
+			array(
+				'nonmin_path'  => 'modules/wordads/js/adflow-loader.js',
+				'dependencies' => array(),
+				'enqueue'      => true,
+				'version'      => JETPACK__VERSION,
+			)
+		);
+
+		wp_enqueue_script(
+			'adflow_config',
+			esc_url( $this->get_config_url() ),
+			array( 'adflow_script_loader' ),
+			JETPACK__VERSION,
+			false
+		);
+
+		$this->is_asset_enqueued = true;
+	}
+
+	/**
+	 * Inserts ad tags on the page.
+	 *
+	 * @return void
+	 */
+	private function insert_ads() {
+		if ( $this->is_amp ) {
+			return;
+		}
+
+		// Don't run on not found pages.
```

Comment:
Do we need to ensure those don't get added to RSS feeds and embeds too, where JS will not be available?

Reply:
For now, I'm going to stick with the same checks that our existing ads are using, but I'll look into this further. The side effect if no JS is available would be an extra invisible `<span>` at the end of the post content—so not too disruptive.